lib/rust/profiler/src/lib.rs

```diff
@@ -431,21 +531,165 @@ pub enum Event {
         /// When the event occurred.
         timestamp: Timestamp,
     },
+    /// Metadata: wrapper with dependency-injected contents.
+    Metadata(Metadata<M>),
+}
+
+impl<M, L> Event<M, L> {
+    /// Produce a new event that may have a different metadata type, with metadata values
+    /// converted by the given function.
+    fn map_metadata<F, M1>(self, mut f: F) -> Event<M1, L>
+    where F: FnMut(M) -> M1 {
+        match self {
+            Event::Start(start) => Event::Start(start),
+            Event::StartPaused(start) => Event::StartPaused(start),
+            Event::End { id, timestamp } => Event::End { id, timestamp },
+            Event::Pause { id, timestamp } => Event::Pause { id, timestamp },
+            Event::Resume { id, timestamp } => Event::Resume { id, timestamp },
+            Event::Metadata(Metadata { timestamp, data }) =>
+                Event::Metadata(Metadata { timestamp, data: f(data) }),
+        }
+    }
```

Comment:
Sometimes it's worth making type ariables longer. I believe this is the case. `M` and `L` and not very descriptive and its hard to understand what they mean. Alternatively, this can be provided in short docs.